pkg/controller/seed-controller-manager/application-secret-cluster-controller/controller.go

```diff
@@ -0,0 +1,258 @@
+
+	"go.uber.org/zap"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1"
+	applicationsecretsynchronizer "k8c.io/kubermatic/v2/pkg/controller/master-controller-manager/application-secret-synchronizer"
+	predicateutil "k8c.io/kubermatic/v2/pkg/controller/util/predicate"
+	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+	"k8c.io/kubermatic/v2/pkg/util/workerlabel"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName  = "kkp-application-secret-cluster-controller"
+	clusterPauseKey = ".spec.pause"
+	isAppSecretKey  = "isAppSecretKey"
+
+	// applicationSecretCleanupFinalizer indicates that secret synced from kubermatic namespace to cluster namespace need cleanup.
+	applicationSecretCleanupFinalizer = "kubermatic.k8c.io/cleanup-application-secret"
+)
+
+type reconciler struct {
+	client              ctrlruntimeclient.Client
+	log                 *zap.SugaredLogger
+	recorder            record.EventRecorder
+	workerName          string
+	namespace           string
+	workerlabelSelector labels.Selector
+}
+
+func Add(
+	ctx context.Context,
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+
+	numWorkers int,
+	workerName string,
+	namespace string,
+
+) error {
+	workerlabelSelector, err := workerlabel.LabelSelector(workerName)
+	if err != nil {
+		return err
+	}
+
+	r := &reconciler{
+		log:                 log.Named(ControllerName),
+		client:              mgr.GetClient(),
+		workerName:          workerName,
+		recorder:            mgr.GetEventRecorderFor(ControllerName),
+		namespace:           namespace,
+		workerlabelSelector: workerlabelSelector,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: r, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestForObject{}, predicateutil.ByAnnotation(applicationsecretsynchronizer.SecretTypeAnnotation, "", false), predicateutil.ByNamespace(r.namespace)); err != nil {
+		return fmt.Errorf("failed to create watch for secrets: %w", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, handler.EnqueueRequestsFromMapFunc(enqueueSecret(ctx, r.client, r.namespace)), workerlabel.Predicates(workerName), noDeleteEventPredicate()); err != nil {
```

Comment:
Wouldn't it make sense to also reconcile clusters that are in deletion?

Reply:
I miss your point...  Application secrets will be deleted at the same time as cluster's namespace. So we can ignore delete events, don't we ?

Reply:
You're right.